### ISSUE TITLE
[PROPOSAL] Neural Search field type

### ISSUE THREAD

**Issue by asfoorial** [opensearch-project/neural-search#803]
Can we mimic this feature in OpenSearch https://www.elastic.co/search-labs/blog/semantic-search-simplified-semantic-text 

I know that a lot has been done recently in OpenSearch projects to make things headache free. I think a neural-search field type in OpenSearch would be an interesting addition. However, it should account for synonyms to avoid any fine-tuning headache.

**Comment by navneet1v** [comment]
@asfoorial from Opensearch side we do this via a combination of ingestion processor and vector field. As there are multiple use-cases for semantic search including multi-model, this would be an interesting field to have.

But is there any specific reason you are looking for the field as compared to what is present currently. My main motive here is to know the advantages of a new field vs what is currently present in opensearch.

**Comment by asfoorial** [comment]
The main reason is simplifying the process and keep the focus on the business. In fact elasticsearch had the same reason when they introduced the field. 

Another reason is alignment of new features across multiple OpenSearch projects. I have noticed over the past number of releases we get new features in ml-commons and kNN. But it takes a while until we see their benefits reflected in neural-search. If they become one component (neural-search field), then that would sort of guarantee that any new feature in ml-common or kNN must be reflected in the neural-search field type before their release.

**Comment by navneet1v** [comment]
> If they become one component (neural-search field), then that would sort of guarantee that any new feature in ml-common or kNN must be reflected in the neural-search field type before their release.

@asfoorial thanks for providing the details. I want to know little bit more on what features added in ML/k-NN doesn't make into Neural. May be there is something missing.

But I really like the idea of having a field which can encapsulate the processor information.

**Comment by navneet1v** [comment]
One place where having the field will be useful is nested fields. I see putting this information in the processor is very painful and not intutive.

**Comment by navneet1v** [comment]
@minalsha please take a look into this and please add your thoughts

**Comment by navneet1v** [comment]
@heemin32 any reason for closing this gh issue?

**Comment by heemin32** [comment]
@navneet1v No. I think it is closed automatically when I added them in NeuralSearch RoadMap.

**Comment by navneet1v** [comment]
One case where I feel this field type will be very useful is in cases of complex nested fields. Currently with TextEmbedding processor it is always feels like we are finding different cases where the processor is not working some GH issues:
1. https://github.com/opensearch-project/neural-search/issues/1042
2. https://github.com/opensearch-project/neural-search/issues/1024
3. #895 
4. #798 
5. #739 
6. #678 
7. #73

I believe having a field type will solve this problem, in the mappers only we will call the MLCommons inference APIs to convert the text to embeddings. I think we can use the concept of properties in the mapper to have a neural field handling both text and vectors. 

cc: @minalsha , @heemin32 , @vibrantvarun , @martin-gaievski

**Comment by YeonghyeonKO** [comment]
This will also reduce the number of inference requests when multiple fields should be embedded. 

> Inference requests in semantic_text fields are also batched. If you have 10 documents in a bulk API request, and each document contains 2 semantic_text fields, then that request will perform a single inference request with 20 texts to your inference service in one go, instead of making 10 separate inference requests of 2 texts each.
(https://www.elastic.co/search-labs/blog/semantic-search-simplified-semantic-text)

**Comment by bzhangam** [comment]
I'll work on this item.

**Comment by YeonghyeonKO** [comment]
@bzhangam, is there room for consideration to include a minor feature? (See: https://github.com/opensearch-project/k-NN/issues/2356)

Either
- Give an warning message about mismatch between original similarity function of embedding model and `space_type` of indices

or 

- Suggest or fix `space_type` when defining mappings for an index according to the embedding model which neural_search field type will use.

**Comment by heemin32** [comment]
@YeonghyeonKO, the `space_type` will be automatically retrieved from the model metadata, so users won't need to specify it explicitly.

**Comment by dblock** [comment]
[[Catch All Triage](https://github.com/opensearch-project/.github/blob/main/TRIAGING.md) - [1](https://github.com/dblock), [2](https://github.com/andrross/), [3](https://github.com/reta), [4](https://github.com/gaiksaya)]

**Comment by mingshl** [comment]
I was having similar idea earlier when I heard about a use case that want to rewrite a match query to neural search query.

Think about this,
- User config the mapping to have a field `text` defined as neural search field, along with a model id, optionally with text chunk size and model config. 
- when the document is ingested, the `text` field will ingest, it auto applies text chunking if needed, and internally call a ml inference processor or text embedding processor, that generates an embedding field call `text_embedding` which holds an array of embeddings. 
- when user are running match query using query text 'foo' to lookup 'text_embedding' field, the query field 'foo' can be rewrite with embedding in a knn query. Or when when running match query with query text 'foo' to lookup 'text_embedding' field, it rewrites to a neural search query. 

This can simplify the neural search experience. But again, we will have to consider how do we handle different model input and output format. For course we can use pre and post processing function through connectors. But what if we can do it easier?

**Comment by bzhangam** [comment]
> This will also reduce the number of inference requests when multiple fields have to be embedded.
> 
> > Inference requests in semantic_text fields are also batched. If you have 10 documents in a bulk API request, and each document contains 2 semantic_text fields, then that request will perform a single inference request with 20 texts to your inference service in one go, instead of making 10 separate inference requests of 2 texts each.
> > (https://www.elastic.co/search-labs/blog/semantic-search-simplified-semantic-text)

I think technically this improvement doesn't have to be done with this new proposal. Even for our existing ingest process we should be able to improve it by overriding this [function](https://github.com/opensearch-project/OpenSearch/blob/6b1861aa93afe382ff4744081f6af98beff506d5/server/src/main/java/org/opensearch/ingest/Processor.java#L96) to pass all the docs to the inferenceProcessor rather than just pass one doc each time.

**Comment by navneet1v** [comment]
> In think if we inference the data in the mapper then each time we will only be able to handle one neural field. In the case we need to inference multiple docs and each doc has multiple neural fields we will invoke the inference API multiple times rather than batching them in one API call. It can be a performance concern especially when we need to use a remote model to do the inference work.

My thoughts:

Doing 1 inference for neural field at a time can for sure decrease the performance compare to batching all the fields and doing inference at once. But at the same time having a field that does the inference solves multiple issues and provides a good cloud native architecture, as system is free from ingest pipelines. Since ingest pipelines require separate node roles and compute power. 

To go around the problem of performance, it will be a good idea to parallelize running of mappers. This will provide a good boost and solve other problems too. This idea has been discussed multiple times in core. I can try to find out references for this. But I have discussed this with couple of maintainers of core. cc: @msfroh

**Comment by bzhangam** [comment]
> This will also reduce the number of inference requests when multiple fields have to be embedded.
> 
> > Inference requests in semantic_text fields are also batched. If you have 10 documents in a bulk API request, and each document contains 2 semantic_text fields, then that request will perform a single inference request with 20 texts to your inference service in one go, instead of making 10 separate inference requests of 2 texts each.
> > (https://www.elastic.co/search-labs/blog/semantic-search-simplified-semantic-text)

I think today we already support a feature for that - https://opensearch.org/docs/latest/ml-commons-plugin/remote-models/batch-ingestion/#step-5-create-an-ingest-pipeline

I think the consideration here is more like with this new proposal we should not lose the ability to batch the inference call.

**Comment by bzhangam** [comment]
> > In think if we inference the data in the mapper then each time we will only be able to handle one neural field. In the case we need to inference multiple docs and each doc has multiple neural fields we will invoke the inference API multiple times rather than batching them in one API call. It can be a performance concern especially when we need to use a remote model to do the inference work.
> 
> My thoughts:
> 
> Doing 1 inference for neural field at a time can for sure decrease the performance compare to batching all the fields and doing inference at once. But at the same time having a field that does the inference solves multiple issues and provides a good cloud native architecture, as system is free from ingest pipelines. Since ingest pipelines require separate node roles and compute power.
> 
> To go around the problem of performance, it will be a good idea to parallelize running of mappers. This will provide a good boost and solve other problems too. This idea has been discussed multiple times in core. I can try to find out references for this. But I have discussed this with couple of maintainers of core. cc: [@msfroh](https://github.com/msfroh)

I agree let the field mapper to do the inference work can simplify things and using parallel can address the latency concern. But we still need to make multiple inference calls. And some ML models may bill us by the number of the API calls which can be a big concern.

**Comment by navneet1v** [comment]
> > > In think if we inference the data in the mapper then each time we will only be able to handle one neural field. In the case we need to inference multiple docs and each doc has multiple neural fields we will invoke the inference API multiple times rather than batching them in one API call. It can be a performance concern especially when we need to use a remote model to do the inference work.
> > 
> > 
> > My thoughts:
> > Doing 1 inference for neural field at a time can for sure decrease the performance compare to batching all the fields and doing inference at once. But at the same time having a field that does the inference solves multiple issues and provides a good cloud native architecture, as system is free from ingest pipelines. Since ingest pipelines require separate node roles and compute power.
> > To go around the problem of performance, it will be a good idea to parallelize running of mappers. This will provide a good boost and solve other problems too. This idea has been discussed multiple times in core. I can try to find out references for this. But I have discussed this with couple of maintainers of core. cc: [@msfroh](https://github.com/msfroh)
> 
> I agree let the field mapper to do the inference work can simplify things and using parallel can address the latency concern. But we still need to make multiple inference calls. And some ML models may bill us by the number of the API calls which can be a big concern.

Some thoughts:

If the concern of making a lot of calls to ML model then user can still use the current ingest pipeline and not use this new field. 

Another option I can think of is putting up a queue which can queue up all the emebedding generation in a batch and then use ML Commons predict api. This way you still get the benefits of mapping field without adding in built extra pipelines.

See if this can help.

**Comment by heemin32** [comment]
I was thinking custom codec for neural search so that batch inference can happen during flush. However, handling it inside mapper might be way simpler to implement. If batch inference is required, we could explore queuing option as @navneet1v mentioned.
